monai/data/grid_dataset.py

```diff
@@ -33,56 +55,83 @@ def __init__(
-            patch to sample from the input arrays. It is assumed the arrays first dimension is the channel dimension which
-            will be yielded in its entirety so this should not be specified in `patch_size`. For example, for an input 3D
-            array with 1 channel of size (1, 20, 20, 20) a regular grid sampling of eight patches (1, 10, 10, 10) would be
-            specified by a `patch_size` of (10, 10, 10).
-
-        """
-        self.patch_size = (None,) + tuple(patch_size)
+        self.patch_size = (None,) + tuple(patch_size)  # expand to have the channel dim
         self.start_pos = ensure_tuple(start_pos)
         self.mode: NumpyPadMode = look_up_option(mode, NumpyPadMode)
         self.pad_opts = pad_opts
 
-    def __call__(self, array):
+    def __call__(self, array: np.ndarray):
         """
         Args:
             array: the image to generate patches from.
+
         """
         yield from iter_patch(
             array,
-            patch_size=self.patch_size,  # expand to have the channel dim
+            patch_size=self.patch_size,  # type: ignore
             start_pos=self.start_pos,
             copy_back=False,
             mode=self.mode,
             **self.pad_opts,
         )
 
 
+class PatchIterd:
+    """
+    Dictionary-based wrapper of :py:class:`monai.data.PatchIter`.
+    Return a patch generator for dictionary data and the coordinate, Typically used
+    with :py:class:`monai.data.GridPatchDataset`.
+    Suppose all the expected fields specified by `keys` have same shape.
+
+    Args:
+        keys: keys of the corresponding items to iterate patches.
+        patch_size: size of patches to generate slices for, 0/None selects whole dimension
+        start_pos: starting position in the array, default is 0 for each dimension
+        mode: {``"constant"``, ``"edge"``, ``"linear_ramp"``, ``"maximum"``, ``"mean"``,
+            ``"median"``, ``"minimum"``, ``"reflect"``, ``"symmetric"``, ``"wrap"``, ``"empty"``}
+            One of the listed string values or a user supplied function. Defaults to ``"wrap"``.
+            See also: https://numpy.org/doc/1.18/reference/generated/numpy.pad.html
+        pad_opts: other arguments for the `np.pad` function.
+            note that `np.pad` treats channel dimension as the first dimension.
+
+    """
+
+    coords_key = "patch_coords"
+
+    def __init__(
+        self,
+        keys: KeysCollection,
+        patch_size: Sequence[int],
+        start_pos: Sequence[int] = (),
+        mode: Union[NumpyPadMode, str] = NumpyPadMode.WRAP,
+        **pad_opts,
+    ):
+        self.keys = ensure_tuple(keys)
+        self.patch_iter = PatchIter(patch_size=patch_size, start_pos=start_pos, mode=mode, **pad_opts)
+
+    def __call__(self, data: Mapping[Hashable, np.ndarray]):
+        d = dict(data)
+        for patch in zip(*[self.patch_iter(d[key]) for key in self.keys]):
+            coords = patch[0][1]  # use the coordinate of the first item
```

Comment:
Are we assuming that every image `d[key]` is the same shape? If so there isn't really a need to consider patch iteration for every image in `[self.patch_iter(d[key]) for key in self.keys]` but only one, so `self.patch_iter(d[first(self.keys)])` assuming `self.keys` is non-empty.

Reply:
Hi @ericspod ,

Thanks for your review.
I don't quite understand your suggestion, could you please help explain more?
I think we need to iterate every image to get the corresponding patches at the same position.

Thanks in advance.

Reply:
Hi @ericspod ,

I think in `{k: v[0] for k, v in zip(self.keys, patch)}`, the `patch` is a list of patches corresponding to every key, like "image", "label", etc. Generated from the `zip(*[XXX])`:
`zip(*[self.patch_iter(d[key]) for key in self.keys])`
So `v` is every item in the list, why you think `v[0]` is always the same?
Or maybe I misunderstood something?

Thanks in advance.

Reply:
Hi @ericspod ,

Thanks for your explanation, I see your point.
I think `self.patch_iter` returns the patch data instead of the slices, so we need to run for every key.
https://github.com/Project-MONAI/MONAI/blob/dev/monai/data/utils.py#L245

Thanks.

Reply:
`self.patch_iter` is giving you both the patch data and the slice indices, if all the images are the same size the patch data will vary but the indices will not. I get what's going on here better and I'm happy for this to go through. We could change some things to maybe make it a little more efficient know the indices are always the same but let's leave that for some other time. Thanks!

Reply:
Thanks for the discussion, I agree with your points.